File: internal/domain/relationship.go
package domain

import (
	"errors"
	"time"
)

// RelationType is IPP-99's fixed enum, kept in sync with the AI service's
// RelationType (services/ai/src/ipp_ai/domain/relationship.py).
type RelationType string

const (
	RelationSupports    RelationType = "supports"
	RelationContradicts RelationType = "contradicts"
	RelationExtends     RelationType = "extends"
	RelationExampleOf   RelationType = "example_of"
	RelationSameTopic   RelationType = "same_topic"
)

func (t RelationType) Valid() bool {
	switch t {
	case RelationSupports, RelationContradicts, RelationExtends, RelationExampleOf, RelationSameTopic:
		return true
	}
	return false
}

var (
	ErrSelfLink             = errors.New("relationship cannot link an insight to itself")
	ErrUnknownRelationType  = errors.New("unknown relation type")
	ErrConfidenceOutOfRange = errors.New("confidence must be within [0,1]")
)

// Relationship is a discovered edge between two insights within a tenant —
// the persisted counterpart to the AI service's in-memory judgement
// (REL 3/IPP-99). REL 4/IPP-100 is what turns it into a stored,
// bidirectional edge (see ports.RelationshipRepository).
type Relationship struct {
	TenantID      string
	FromInsightID string
	ToInsightID   string
	Type          RelationType
	Confidence    float64
	Rationale     string
	DiscoveredAt  time.Time
}

// Validate checks the fields that don't require a database round trip:
// self-links, the relation type enum, and the confidence range. Whether
// FromInsightID/ToInsightID actually exist in the tenant is
// RelationshipRepository.Put's job — only it can check that.
func (r Relationship) Validate() error {
	if r.FromInsightID == r.ToInsightID {
		return ErrSelfLink
	}
	if !r.Type.Valid() {
		return ErrUnknownRelationType
	}
	if r.Confidence < 0 || r.Confidence > 1 {
		return ErrConfidenceOutOfRange
	}
	return nil
}

File: services/ai/src/ipp_ai/domain/relationship.py
"""Relationship judgement — pure value types, no I/O.

The read side's counterpart to internal/domain.Relationship (IPP-100),
minus the fields only persistence needs (`DiscoveredAt`, the deterministic
`sk`): this is what REL 3 produces in memory, one candidate pair at a time.
REL 4 is what turns a Relationship into a stored, bidirectional edge.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum


class RelationType(StrEnum):
    """IPP-99's fixed enum. A value outside this set is not a RelationType —
    construct via `RelationType(raw)` and let the ValueError propagate;
    there is no `.get()`-style coercion path from an unrecognized string.
    """

    SUPPORTS = "supports"
    CONTRADICTS = "contradicts"
    EXTENDS = "extends"
    EXAMPLE_OF = "example_of"
    SAME_TOPIC = "same_topic"


@dataclass(frozen=True)
class RelationJudgement:
    """One LLM call's verdict on a single candidate pair, before the
    confidence threshold (application/relationship.py) decides whether it
    becomes a Relationship.
    """

    relation_type: RelationType
    confidence: float
    rationale: str


@dataclass(frozen=True)
class Relationship:
    """A judgement that cleared the confidence threshold, addressed to a
    specific ordered pair of insights.
    """

    tenant_id: str
    from_insight_id: str
    to_insight_id: str
    relation_type: RelationType
    confidence: float
    rationale: str


@dataclass(frozen=True)
class RelatedInsight:
    """One edge from an insight's perspective — the read side's counterpart
    to internal/domain.RelatedInsight (REL 6/IPP-102). `insight_id` is
    whichever insight is on the *other* end from the one you queried; `text`
    is denormalized onto the edge at write time (see the Go adapter), so
    reading it never needs a second fetch.
    """

    insight_id: str
    text: str
    relation_type: RelationType
    confidence: float
    rationale: str
